cxx/pclean/pclean_lib.cc

```diff
@@ -58,3 +59,63 @@ T_observations translate_observations(
   }
   return obs;
 }
+
+// Sample a single "row".
+void make_pclean_sample(
+    HIRM *hirm, const PCleanSchema& schema,
+    const std::map<std::string, std::vector<std::string>>
+    &annotated_domains_for_relations,
+    std::mt19937* prng,
+    std::map<std::string, std::string> *query_values) {
+  std::map<std::string, CRP> domain_crps;
+  hirm->initialize_domain_crps(&domain_crps);
+
+  // entity_assignments[annotated_entity] gives the entity id for that entity.
+  std::map<std::string, int> entity_assignments;
+  for (const auto& query_field : schema.query.fields) {
+    T_items entities;
+    const std::vector<std::string>& domains = std::visit(
+        [](auto trel) { return trel.domains; },
+        hirm->schema[query_field.name]);
+    const std::vector<std::string>& annotated_domains =
+        annotated_domains_for_relations.at(query_field.name);
+    if (domains.size() != annotated_domains.size()) {
+      printf("For relation %s, found %ld domains but %ld annotated domains\n",
+             query_field.name.c_str(), domains.size(), annotated_domains.size());
+      std::exit(1);
+    }
+    for (size_t i = 0; i < domains.size(); ++i) {
+      int id = -1;
+      auto it = entity_assignments.find(annotated_domains[i]);
```

Comment:
It looks like at each call to `make_pclean_sample` the entity assignments are sampled independently, is that right? What I mean is if we've already sampled `physician:Physician == 3, physician:school:School == 5` in some previous sample, in a future sample we can get `physician:Physician == 3, physician:school:School == 2`, which shouldn't be possible because the first sample implies that Physician 3 went to School 5.

(If that's the case, instead of fixing it here we could punt on it until Model 7, since I'm having to deal with that in Model 7 too.)

Reply:
Yes, the entity assignments are recreated for every row.  I added a TODO to fix that.

Reply:
I would prefer not to check this in until it's fixed, since the method currently returns invalid samples (or at least to highlight it more, maybe by renaming the function something that indicates that samples are invalid and it is a WIP)

Reply:
Please suggest a name you would be happy with, and I'll change it to that.

Reply:
Up to you, maybe just append WIP_

Reply:
prepend i mean

Reply:
Done.